### ISSUE TITLE
Not timing out and going into an infinite loop when multiplexed connection is terminated

### ISSUE THREAD

**Issue by sunilvirus** [curl/curl#4135]
### I did this
I use multi socket API. I was using libcurl for HTTP v2 with multiplexing for some time and my code was working fine with 7.62.0 (albeit with unnecessary timeout errors because of the bugs that got fixed in later version of the library). I did not make any code changes in my application code, but just upgraded my libraries to 7.65.2. 

My test is very simple. My code will attempt to send multiple post requests in parallel and I keep restarting the webserver. 

### I expected the following
I am expecting the callback to be called with a timeout for the request or finish the request. But some requests are getting into an infinite loop.

### libcurl version
7.65.2

### Operating System
Ubuntu 18

I am not sure if I am doing something wrong and will be happy to be proven wrong with an explanation. The symptom as I said is that some (multiplexed HTTP v2) requests are not timing out and I can see the execution is going into an infinite loop. I am digging into the code for the last 3 days and I would like to know.

Consider the following situation in `multi_runsingle()`.
There are two pending requests on a multiplexed connection.
And `data->mstate` is `CURLM_STATE_DO` for both of them.
When one of the easy connection goes to case `CURLM_STATE_DO`, it does `multi_do()` and fails with `CURLE_SEND_ERROR`.

In `multi_done()`, `detach_connnection()` is called but returns from there because `CONN_INUSE()` is true;
After `multi_done()`, the following lines of code will be executed in `multi_runsingle()`
```
 multistate(data, CURLM_STATE_CONNECT);
 rc = CURLM_CALL_MULTI_PERFORM;
 result = CURLE_OK;
```
by the end of statemachine
```
data->conn is NULL (because of detach_connnection())
result is CURLE_OK
rc is CURLM_CALL_MULTI_PERFORM
data->mstate is CURLM_STATE_CONNECT
```
So, the loop continues without checking for timeout etc towards the end of the loop.
Again, before entering the state machine, it wont check timeout because `data->conn == NULL`

In the state machine, it goes to `CURLM_CALL_MULTI_PERFORM `(obviously)
In `Curl_connect()`, it will find the existing (bad) connection going to the same URL.
It thinks that it can multiplex on that and attaches to the connection.
Changes state to `CURLM_STATE_DO` (assume non-async for sake of simplicity)

Now the whole cycle repeats infinitely. I could see the continuous detach and attach activity (with my extra diagnostic code). Is there a breaking condition that is supposed to trigger and my misuse of API is causing this problem ? Here are the debug logs to show the flow that I described above.
```
* Found bundle for host n1.sunil.local: 0x7f8227e0fbc8 [can multiplex]
* Multiplexed connection found!
* Found connection 1, with 1 requests on it
* Re-using existing connection! (#1) with host n1.sunil.local
* Transfer was pending, now try another
* STATE: CONNECT => DO handle 0x7f8227e7f008; line 1407 (connection #1)
* Send failure: Bad file descriptor
* Failed sending HTTP POST request
* Connection died, retrying a fresh connect
* multi_done
* Connection still in use 1, no more multi_done now!
* Issue another request to this URL: 'http://n1.sunil.local:2101/sunil.html'
```

**Comment by bagder** [comment]
Possibly a duplicate of or similar to #4088

**Comment by ngg** [comment]
I have probably the same issue but without using HTTP/2. We also upgraded from 7.62.0 to 7.65.2 and something broke, some of our connections are not terminating after some network errors.
We're using the multi interface (the problem happens both using the curl_multi_wait api and with an external event loop implementation), we have multiple easy connections in the same multi, lots of those are connecting to the same servers, connections can be reused really often.
Unfortunately I couldn't yet reliably reproduce this, but I'm trying to create a simple example that has this issue. I'll try to git bisect the issue once I can reproduce more reliably.

**Comment by sunilvirus** [comment]
> Possibly a duplicate of or similar to #4088

In case of #4088, the error they are getting is EAGAIN on `recvfrom`() call. That means the fd is ok and it is the right thing to continue on an EAGAIN. I get EBADF on attempt to do write(). So, the same connection should not be used. And and important difference is that I do not get timeout error when I am expecting it, whereas in case of #4088 they are getting timeout. 

One more thing I observed is that when the control goes back to `CURLM_STATE_CONNECT `in statemachine, we call  `Curl_pgrsTime(data, TIMER_STARTSINGLE);`. This gives a fresh lease of life for the request as if it is starting afresh. So, even if we do a timeout check, it will not be a timeout. 

Any comments/insights on my original analysis ? Either to support it or refute it ?

**Comment by ngg** [comment]
I came up with the following modified version of the multi-single.c example code.

It has a single multi interface (with total num of connections limited to 3), an infinite loop, and an infinite amount of coming new easy connections (it always fills up its queue to 10). When running this, I'm turning on and off my ethernet interface, and see what happens.

I think this should "work", which means that even after long connection losses, it should continue downloading after the network is repaired.

There is a behaviour difference between 7.62.0 and 7.65.1, it looks like newer curls don't handle well something in the connection pool after DNS errors (after some connections fail with CURLE_COULDNT_RESOLVE_HOST). The easiest way to reproduce is to disable the internet connection before starting the example, older curl versions write lots of errors (as lots of connections fail), newer versions only report 3 errors and then hang. If I enable the internet later, older versions continue transfers, newer ones are still hanging.

I've bisected the issue to https://github.com/curl/curl/commit/2f44e94efb3df8e50bb2ddbc4ec6b569a6424517 (the removal of HTTP pipelining, https://github.com/curl/curl/pull/3651), that commit fails, and the one before that was working. I'm not sure this is the same issue as the one reported here originally, but it's really similar.

```c
#include <curl/curl.h>
#include <stdio.h>
#include <string.h>
#include <sys/time.h>
#include <unistd.h>

#define WAITMS(x)                          \
    struct timeval wait = { 0, (x)*1000 }; \
    (void)select(0, NULL, NULL, NULL, &wait);

int main(void)
{
    int still_running = 0;
    int last = 0;
    int repeats = 0;
    int num_easy = 0;
    FILE* devnull = fopen("/dev/null", "w");
    curl_global_init(CURL_GLOBAL_DEFAULT);

    CURLM* multi_handle = curl_multi_init();
    curl_multi_setopt(multi_handle, CURLMOPT_MAX_TOTAL_CONNECTIONS, 3L);

    for (;;) {
        for (; num_easy < 10; num_easy++) {
            // Fill up easy queue, we should always have 10 active or waiting
            CURL* http_handle = curl_easy_init();
            curl_easy_setopt(http_handle, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
            curl_easy_setopt(http_handle, CURLOPT_URL, "http://code.jquery.com/jquery-3.4.1.js"); // random not so small file from a CDN
            curl_easy_setopt(http_handle, CURLOPT_WRITEDATA, (void*)devnull);
            curl_easy_setopt(http_handle, CURLOPT_CONNECTTIMEOUT, 5L);
            curl_easy_setopt(http_handle, CURLOPT_LOW_SPEED_LIMIT, 10L);
            curl_easy_setopt(http_handle, CURLOPT_LOW_SPEED_TIME, 5L);
            curl_multi_add_handle(multi_handle, http_handle);
        }

        CURLMcode mc = curl_multi_perform(multi_handle, &still_running);
        if (mc != CURLM_OK) {
            fprintf(stderr, "curl_multi_perform() failed, code %d.\n", mc);
            return 1;
        }

        int pending;
        CURLMsg* msg;
        while ((msg = curl_multi_info_read(multi_handle, &pending)) != NULL) {
            if (msg->msg != CURLMSG_DONE) {
                fprintf(stderr, "unknown curlmsg %d\n", msg->msg);
                return 1;
            }
            fprintf(stderr, "result = %d (%s)\n", msg->data.result, curl_easy_strerror(msg->data.result));
            CURL* e = msg->easy_handle;
            curl_multi_remove_handle(multi_handle, e);
            curl_easy_cleanup(e);
            num_easy--;
        }

        if (still_running != last) {
            fprintf(stderr, "still_running = %d\n", still_running);
            last = still_running;
        }

        int numfds;
        mc = curl_multi_wait(multi_handle, NULL, 0, 1000, &numfds);
        if (mc != CURLM_OK) {
            fprintf(stderr, "curl_multi_wait() failed, code %d.\n", mc);
            return 1;
        }

        if (!numfds) {
            repeats++;
            if (repeats > 1) {
                WAITMS(100);
            }
        } else {
            repeats = 0;
        }
    }
}
```

**Comment by ngg** [comment]
@sunilvirus can you please check if the first commit that's not working for you is https://github.com/curl/curl/commit/2f44e94efb3df8e50bb2ddbc4ec6b569a6424517 or something else?

**Comment by sunilvirus** [comment]
> @sunilvirus can you please check if the first commit that's not working for you is [2f44e94](https://github.com/curl/curl/commit/2f44e94efb3df8e50bb2ddbc4ec6b569a6424517) or something else?

@ngg , seems you are on to something. I tested with one commit before [2f44e94] (i.e [aba1c51]) and I do not hit the same infinite loop problem. However, after long time into my testing, I am hitting #3986 (which is probably expected). But with [2f44e94], I hit the infinite loop problem in just < 5 iterations.

**Comment by kunalekawde** [comment]
@sunilvirus , could please check if the issue is similar to #3991 ? I had made some changes to application code as well as noted in there apart from lib fixes.

**Comment by sunilvirus** [comment]
@ngg , to add more observations. I do not get into infinite loop when using HTTPv1 with [2f44e94] or even the latest and greatest tag 'curl-7_65_3' i.e [aa73eb4]. I ran about 50 iterations of my test without getting into this infinite loop when using HTTPv1 while I hit the infinite loop in < 5 iterations with HTTPv2 & multiplexing.

Like I said in my original analysis, a precondition is that two (or more) easy handles should be sharing a connection. If there is only one easy handle, when `multi_done()` is called, the code proceeds further after `detach_connection()` as `CONN_INUSE()` will be false. It will go on to call `Curl_disconnect()` as `conn->bits.close` is set to `true `in `Curl_retry_request()->connclose()->Curl_conncontrol()`

**Comment by ngg** [comment]
I have an even simpler example for my case (I'll soon try to check if it's doing the same thing as in yours), this example runs infinitely since https://github.com/curl/curl/commit/2f44e94efb3df8e50bb2ddbc4ec6b569a6424517 and was working before that. Currently only the first connection finishes with error and the second hangs.
```c
#include <curl/curl.h>
#include <stdio.h>

int main(void)
{
    curl_global_init(CURL_GLOBAL_DEFAULT);

    CURLM* multi_handle = curl_multi_init();
    curl_multi_setopt(multi_handle, CURLMOPT_MAX_TOTAL_CONNECTIONS, 1L);

    for (int i = 0; i < 2; i++) {
        CURL* http_handle = curl_easy_init();
        curl_easy_setopt(http_handle, CURLOPT_URL, "http://non-existing-host.haxx.se");
        curl_multi_add_handle(multi_handle, http_handle);
    }

    int still_running, last = 0;
    do {
        CURLMcode mc = curl_multi_perform(multi_handle, &still_running);
        if (mc != CURLM_OK) {
            fprintf(stderr, "curl_multi_perform() failed, code %d.\n", mc);
            return 1;
        }
        if (still_running != last) {
            fprintf(stderr, "still_running = %d\n", still_running);
            last = still_running;
        }

        int numfds;
        mc = curl_multi_wait(multi_handle, NULL, 0, 1000, &numfds);
        if (mc != CURLM_OK) {
            fprintf(stderr, "curl_multi_wait() failed, code %d.\n", mc);
            return 1;
        }
    } while (still_running);
}
```

**Comment by sunilvirus** [comment]
@ngg , in this latest code, you are not setting the easy option CURLOPT_HTTP_VERSION. This will default to HTTPv2. Yes, this issue almost readily happens with HTTPv2 in my experiment.

**Comment by ngg** [comment]
@sunilvirus HTTP2 is only default for TLS, it's simple HTTP which uses 1.1 by default. But it doesn't matter because DNS resolution will fail. It fails the same way if I set it to CURL_HTTP_VERSION_1_1 or CURL_HTTP_VERSION_2_0.

**Comment by sunilvirus** [comment]
@kunalekawde , I can see some similarities with #3991. The call stack is same in my case too. Your debug logs are also pretty close to mine. So, they look pretty close if not exactly the same. But you closed your issue saying that #3997 fixed you issue. I can see this infinite loop in 7.65.2 which includes the fix.

**Comment by ngg** [comment]
I think at the end these are two separate issues (both probably caused by #3651), so I opened a new issue with my case #4144

**Comment by ngg** [comment]
@sunilvirus Can you please check if https://github.com/curl/curl/pull/4151 solves your issue? It seems to solve mine.

**Comment by sunilvirus** [comment]
In my issue, the code does not flow through this path. As I mentioned in my original analysis, by the end of state machine, the 'result' is CURLE_OK. Also, data->conn is also false as `detach_connection()` is done in `multi_done()`. So, the code flow will not even enter the if condition which #4151 is fixing.  Nevertheless, I will test it to be sure.

**Comment by bagder** [comment]
#4151 has been merged into master so it should be an easy test to see if the bug remains...

**Comment by sunilvirus** [comment]
@bagder / @ngg, No luck with the latest master for this issue. I still can reproduce this.

**Comment by bagder** [comment]
@sunilvirus as far as I can see, you haven't provided any sample code or anything to allow us to reproduce this problem so I'm a little curious what you're expecting from us here.

@ngg provided two code snippets above, but it sounds to me that they're not showing your problem?

I cannot reproduce this problem!

**Comment by sunilvirus** [comment]
Sorry for the delayed response. I had to switch to HTTPv1 for the time being. I will try to come up with an independent code to have a repro outside my code.

**Comment by bagder** [comment]
Without more info to go on, I will feel obliged to close this issue soon - then you can reopen it again or create a new one if you at a later point can provide a recipe we can reproduce with.

**Comment by sunilvirus** [comment]
Ok. Please close it for now. I had to move of the the machine where I was able to readily reproduce it (with my code). I tried it with a fresh setup and I am not able to repro this. I am not sure what changed in the new setup. I will get back to you if I can find a reliable and independent repro.

**Comment by bagder** [comment]
Unable to reproduce.
